core/modules/gtpu_encap.cc

```diff
@@ -76,187 +66,42 @@ struct [[gnu::packed]] PacketTemplate {
-      std::cerr << "IP Address: " << ToIpv4Address(be32_t(ip))
-                << ", Data: " << data << std::endl;
-    }
-  } while (rc >= 0);
-
-  return CommandSuccess();
-}
-/*----------------------------------------------------------------------------------*/
-CommandResponse GtpuEncap::ShowCount(const bess::pb::EmptyArg &) {
-  bess::pb::GtpuEncapArg arg;
-  arg.set_s1u_sgw_ip(0);
-  arg.set_num_subscribers(rte_hash_count(session_map));
-  DLOG(INFO) << "# of records: " << rte_hash_count(session_map) << std::endl;
-  return CommandSuccess(arg);
-}
-/*----------------------------------------------------------------------------------*/
 void GtpuEncap::ProcessBatch(Context *ctx, bess::PacketBatch *batch) {
   int cnt = batch->cnt();
-  int hits = 0;
-  uint64_t key[bess::PacketBatch::kMaxBurst];
-  void *key_ptr[bess::PacketBatch::kMaxBurst];
-  struct session_info *data[bess::PacketBatch::kMaxBurst];
-  uint64_t hit_mask = 0ULL;
 
   for (int i = 0; i < cnt; i++) {
     bess::Packet *p = batch->pkts()[i];
-    /* assuming that this module comes right after EthernetDecap */
-    /* pkt_len can be used as the length of IP datagram */
-    Ipv4 *iph = p->head_data<Ipv4 *>();
-    be32_t daddr = iph->dst;
-    be32_t saddr = iph->src;
-    DLOG(INFO) << "ip->saddr: " << ToIpv4Address(saddr)
-               << ", ip->daddr: " << ToIpv4Address(daddr) << std::endl;
-    key[i] = SESS_ID(daddr.raw_value(), DEFAULT_BEARER);
-    key_ptr[i] = &key[i];
-  }
 
-  if ((hits = rte_hash_lookup_bulk_data(session_map, (const void **)&key_ptr,
-                                        cnt, &hit_mask, (void **)data)) <= 0) {
-    DLOG(INFO) << "Failed to look-up" << std::endl;
-    /* Since default module is sink, the packets go right in the dump */
-    /* RunNextModule() sends batch to DEFAULT GATE */
-    RunNextModule(ctx, batch);
-    return;
-  }
-
-  DLOG(INFO) << "rte_hash_lookup_bulk_data output: (cnts: " << cnt
-             << ", hits: " << hits << ", hit_mask: " << hit_mask << ")"
-             << std::endl;
-
-  for (int i = 0, j = 0; i < cnt && j < hits; i++) {
-    bess::Packet *p = batch->pkts()[i];
-    if (!ISSET_BIT(hit_mask, i)) {
-      EmitPacket(ctx, p, DEFAULT_GATE);
-      DLOG(INFO) << "Fetch failed for ip->daddr: "
-                 << ToIpv4Address(be32_t(UE_ADDR(key[i]))) << std::endl;
-      continue;
-    }
+    /* check attributes' values now */
+    uint32_t at_tout_sip;
+    bess::metadata::mt_offset_t off = attr_offset(tout_sip_attr);
+    at_tout_sip = get_attr_with_offset<uint32_t>(off, p);
+
+    uint32_t at_tout_dip;
+    off = attr_offset(tout_dip_attr);
+    at_tout_dip = get_attr_with_offset<uint32_t>(off, p);
+
+    uint32_t at_tout_teid;
+    off = attr_offset(tout_teid);
+    at_tout_teid = get_attr_with_offset<uint32_t>(off, p);
+
+    uint16_t at_tout_uport;
+    off = attr_offset(tout_uport);
+    at_tout_uport = get_attr_with_offset<uint16_t>(off, p);
```

Comment:
ether encap style? and remove tout_sip_attr, tout_dip_attr, tout_teid, tout_uport?

```suggestion
    uint32_t tun_src;
    uint32_t tun_dst;
    uint32_t tun_eid;
    uint16_t tun_port;

    tun_src = get_attr<uint32_t>(this, ATTR_R_TUN_SRC, p);
    tun_dst = get_attr<uint32_t>(this, ATTR_R_TUN_DST, p);
    tun_eid = get_attr<uint32_t>(this, ATTR_R_TUN_EID, p);
    tun_port = get_attr<uint16_t>(this, ATTR_R_TUN_PORT, p);
```

Reply:
I would prefer it to keep it this way. This results in less number of macro declarations.